File: android/common/data/src/main/java/com/jaychang/food/data/SessionRepository.kt
/*
 * Copyright (C) 2021. Jay Chang
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.jaychang.food.data

import com.jaychang.food.api.apis.SessionApi
import com.jaychang.food.api.models.Session
import com.jaychang.food.api.models.SessionRequestBody
import com.jaychang.food.api.toApiError
import com.jaychang.food.data.SessionLoginResult.Error.InvalidPassword
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

sealed class SessionLoginResult {
    data class Success(val data: Session) : SessionLoginResult()
    sealed class Error : SessionLoginResult() {
        object InvalidPassword : Error() {
            const val code = "user_invalid_password"
        }
    }
}

@Singleton
class SessionRepository @Inject constructor(
    private val sessionApi: SessionApi,
    private val sessionManager: SessionManager
) {
    fun login(phoneNumber: String, password: String): Single<SessionLoginResult> {
        return sessionApi.postSessions(SessionRequestBody(phoneNumber, password))
            .doOnSuccess { sessionManager.saveSession(it)  }
            .map<SessionLoginResult> { SessionLoginResult.Success(it) }
            .onErrorReturn {
                if (it.toApiError()?.code == InvalidPassword.code) {
                    InvalidPassword
                } else {
                    throw it
                }
            }
            .subscribeOn(Schedulers.io())
    }

    fun isLoggedIn(): Single<Boolean> = sessionManager.isLoggedIn()
}

File: flutter/food/common/data/lib/src/session_repository.dart
import 'package:api/api.dart';
import 'session_manager.dart';

class SessionRepository {
  SessionRepository({
    required sessionApi,
    required sessionManager
  }) : _sessionApi = sessionApi,
       _sessionManager = sessionManager;

  final SessionApi _sessionApi;

  final SessionManager _sessionManager;

  Future<SessionLoginResult> login(String phoneNumber, String password) async {
    final body = SessionRequestBody((b) {
      b.phoneNumber = phoneNumber;
      b.password = password;
    });
    try {
      final response = await _sessionApi.postSessions(sessionRequestBody: body);
      final session = response.data!;
      _sessionManager.saveSession(session);
      return SessionLoginResultSuccess(session);
    } on Exception catch (e) {
      final apiError = e.toApiError();
      if (apiError?.code == "user_invalid_password") {
        return SessionLoginResultInvalidPassword();
      }
      rethrow;
    }
  }

  Future<bool> isLoggedIn() => _sessionManager.isLoggedIn();
}

abstract class SessionLoginResult {}

class SessionLoginResultSuccess extends SessionLoginResult {
  SessionLoginResultSuccess(this.data);

  final Session data;
}

class SessionLoginResultInvalidPassword extends SessionLoginResult {}
